### PR TITLE
[REVIEW] Fix a few of the Cython warnings

### PR DESCRIPTION
**Pull request by trxcllnt**
Partially addresses https://github.com/rapidsai/cudf/issues/4024

Fixes these classes of warnings:
```
* Use boundscheck(False) for faster access
* local variable 'c_category' referenced before assignment
* Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release!
```

Doesn't fix these:
```
* Not all members given for struct 'reader_options'
* Not all members given for struct 'csv_write_arg'
```

This PR includes commit in https://github.com/rapidsai/cudf/pull/4222 and adds an optional `BUILD_LEGACY_TESTS` flag to the cmake scripts. This flag is opt-in, and will skip building/linking any C++ test of the pattern `LEGACY_<name>_TEST`. Having this flag on makes libcudf C++ build faster.

These are here because I'm trying to track down the debug-build linker issue, and I have to be careful about switching branches while the build is in progress. I'd be happy to cherry-pick the cmake change into a separate PR after this if necessary.

### CODE REVIEW

File: python/cudf/setup.py [hunk 1]
```diff
@@ -77,7 +77,7 @@
     ],
     # Include the separately-compiled shared library
     setup_requires=["cython"],
-    ext_modules=cythonize(extensions, nthreads=nthreads),
+    ext_modules=cythonize(extensions, nthreads=nthreads, language_level="3"),
```
Comment (kkraus14) [inline python/cudf/setup.py]: Can we move all of the compiler directives to here? I.e. `profile=False`, `language=c++`, `embedsignature=True`, `boundscheck=False`? If so we can get rid of the header from all of our Cython files and not worry about it moving forward 😄

File: cpp/tests/CMakeLists.txt [hunk 2]
```diff
@@ -44,15 +44,20 @@ install(TARGETS cudftestutil
...
+        add_executable(${CMAKE_TEST_NAME}
+                       ${CMAKE_TEST_SRC})
+        set_target_properties(${CMAKE_TEST_NAME} PROPERTIES POSITION_INDEPENDENT_CODE ON)
+        target_link_libraries(${CMAKE_TEST_NAME} gmock gtest gmock_main gtest_main pthread cudf cudftestutil)
+        set_target_properties(${CMAKE_TEST_NAME} PROPERTIES
+                                RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/gtests")
+        add_test(NAME ${CMAKE_TEST_NAME} COMMAND ${CMAKE_TEST_NAME})
+        set(CUDF_TEST_LIST ${CUDF_TEST_LIST} ${CMAKE_TEST_NAME} CACHE INTERNAL "CUDF_TEST_LIST")
+    endif(BUILD_LEGACY_TESTS OR (NOT is_legacy_test))
 endfunction(ConfigureTest)
```
Comment (kkraus14) [inline cpp/tests/CMakeLists.txt]: This feels out of scope for this PR and should be a separate PR

File: python/cudf/setup.py (hunk 1 above)
Comment (trxcllnt) [reply python/cudf/setup.py]: Yes we can, but be careful to only include ones that should apply globally. I'm not sure if `boundscheck=False` should apply everywhere or not, since the warning only showed up in the `io` modules.

Comment (kkraus14) [reply python/cudf/setup.py]: Okay, well how about lets do it for everything other than `boundscheck` for now?

Comment (trxcllnt) [reply python/cudf/setup.py]: Yep, that sounds like a good idea.

File: cpp/tests/CMakeLists.txt (hunk 2 above)
Comment (trxcllnt) [reply cpp/tests/CMakeLists.txt]: Yes, I was in the middle of a debug build and couldn't switch branches, but also didn't want to lose this change. I'll cherry-pick this commit into a separate branch. The easiest thing after that is to leave this commit in here, but merge them independently.

File: python/cudf/cudf/_lib/dlpack.pyx [hunk 3]
```diff
@@ -1,11 +1,6 @@
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
 
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
-
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
```
Comment (cwharris) [inline python/cudf/cudf/_lib/dlpack.pyx]: duplicate

File: python/cudf/cudf/_lib/groupby.pyx [hunk 4]
```diff
@@ -1,11 +1,6 @@
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
 
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
-
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
```
Comment (cwharris) [inline python/cudf/cudf/_lib/groupby.pyx]: duplicate

File: python/cudf/cudf/_lib/includes/search.pxd [hunk 5]
```diff
@@ -1,11 +1,6 @@
-# Copyright (c) 2019, NVIDIA CORPORATION.
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
 
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
-
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
```
Comment (cwharris) [inline python/cudf/cudf/_lib/includes/search.pxd]: duplicate

File: python/cudf/cudf/_lib/issorted.pyx [hunk 6]
```diff
@@ -1,11 +1,6 @@
-# Copyright (c) 2019, NVIDIA CORPORATION.
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
 
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
-
-# Copyright (c) 2019, NVIDIA CORPORATION.
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
```
Comment (cwharris) [inline python/cudf/cudf/_lib/issorted.pyx]: duplicate

File: python/cudf/cudf/_lib/nvtx.pyx [hunk 7]
```diff
@@ -1,11 +1,6 @@
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
 
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
-
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
```
Comment (cwharris) [inline python/cudf/cudf/_lib/nvtx.pyx]: duplicate

File: python/cudf/cudf/_lib/quantile.pyx [hunk 8]
```diff
@@ -1,11 +1,6 @@
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
 
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
-
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
```
Comment (cwharris) [inline python/cudf/cudf/_lib/quantile.pyx]: duplicate

File: python/cudf/cudf/_lib/reduce.pyx [hunk 9]
```diff
@@ -1,11 +1,6 @@
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
 
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
-
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
```
Comment (cwharris) [inline python/cudf/cudf/_lib/reduce.pyx]: duplicate

File: python/cudf/cudf/_lib/replace.pyx [hunk 10]
```diff
@@ -1,11 +1,6 @@
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
 
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
-
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
```
Comment (cwharris) [inline python/cudf/cudf/_lib/replace.pyx]: duplicate

File: python/cudf/cudf/_lib/search.pyx [hunk 11]
```diff
@@ -1,11 +1,6 @@
-# Copyright (c) 2019, NVIDIA CORPORATION.
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
 
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
-
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
```
Comment (cwharris) [inline python/cudf/cudf/_lib/search.pyx]: duplicate

File: python/cudf/cudf/_lib/sort.pyx [hunk 12]
```diff
@@ -1,11 +1,6 @@
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
 
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
-
-# Copyright (c) 2018, NVIDIA CORPORATION.
+# Copyright (c) 2018-2020, NVIDIA CORPORATION.
```
Comment (cwharris) [inline python/cudf/cudf/_lib/sort.pyx]: duplicate

File: python/cudf/cudf/_lib/avro.pyx [hunk 13]
```diff
@@ -1,9 +1,6 @@
-# Copyright (c) 2019, NVIDIA CORPORATION.
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+# cython: boundscheck = False
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/avro.pyx]:
```suggestion

# cython: boundscheck = False
```

File: python/cudf/cudf/_lib/csv.pyx [hunk 14]
```diff
@@ -1,9 +1,5 @@
-# Copyright (c) 2019, NVIDIA CORPORATION.
-
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+# cython: boundscheck = False
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/csv.pyx]:
```suggestion

# cython: boundscheck = False
```

File: python/cudf/cudf/_lib/json.pyx [hunk 15]
```diff
@@ -1,9 +1,5 @@
-# Copyright (c) 2019, NVIDIA CORPORATION.
-
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+# cython: boundscheck = False
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/json.pyx]:
```suggestion

# cython: boundscheck = False
```

File: python/cudf/cudf/_lib/orc.pyx [hunk 16]
```diff
@@ -1,9 +1,5 @@
-# Copyright (c) 2019, NVIDIA CORPORATION.
-
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+# cython: boundscheck = False
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/orc.pyx]:
```suggestion

# cython: boundscheck = False
```

File: python/cudf/cudf/_lib/parquet.pyx [hunk 17]
```diff
@@ -1,9 +1,5 @@
-# Copyright (c) 2019, NVIDIA CORPORATION.
-
-# cython: profile=False
-# distutils: language = c++
-# cython: embedsignature = True
-# cython: language_level = 3
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+# cython: boundscheck = False
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/parquet.pyx]:
```suggestion

# cython: boundscheck = False
```

**Comment by cwharris** [comment]
Oh, it looks like those dups were in there already. :man_shrugging:

File: python/cudf/setup.py (hunk 1 above)
Comment (trxcllnt) [reply python/cudf/setup.py]: image

Do the recent commits resolve this, or are there other compiler directives we should include?

**Comment by jakirkham** [comment]
Thanks for working on this @trxcllnt! 😄
